### PR TITLE
Return correct loginname in credentials

### LINKED ISSUE #21285: In LDAP setup (uid != loginname), token created through getapppassword cease to function after a few minutes
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. curl -u loginname:mainpawssord -H 'OCS-APIRequest: true' https://cloud.example.com/ocs/v2.php/core/getapppassword
2. this returns a new app token APPPWD
3. APPPWD is usable with UID, not with loginname. E.g.:
curl -T textfile -u uid:APPPWD https://cloud.example.com/remote.php/webdav/path/to/directory/
works, but
curl -T textfile -u loginname:APPPWD https://cloud.example.com/remote.php/webdav/path/to/directory/
does not
4. after a few minutes, it is not usable at all. That is:
sleep 600 && curl -T textfile -u uid:APPPWD https://cloud.example.com/remote.php/webdav/path/to/directory/
does not work (gives an authentication error), and from that point on
curl -T textfile -u uid:APPPWD https://cloud.example.com/remote.php/webdav/path/to/directory/
never works again.

### Expected behaviour
The app token should work indefinitely until revoked by the user, or the user is disabled / deleted, or some similar event.

### Actual behaviour
The app token stops to work after a few minutes.

### Server configuration

**Operating system:** Debian GNU/Linux

**Web server:** Apache2

**Database:** PostgreSQL

**PHP version:** 7.3.14

**Nextcloud version:** 17.0.7

**Updated from an older Nextcloud/ownCloud or fresh install:** update from 17.0.2, itself updated from probably 17.0.0

**Where did you install Nextcloud from:** downloaded zip from https://download.nextcloud.com/server/releases/

**Signing status:** irrelevant, since I fixed the bug (changed the code) and this obviously makes signing errors

**List of activated apps:**
Enabled:
  - accessibility: 1.3.0
  - activity: 2.10.1
  - cloud_federation_api: 1.0.0
  - comments: 1.7.0
  - dav: 1.13.0
  - extract: 1.2.3
  - federatedfilesharing: 1.7.0
  - federation: 1.7.0
  - files: 1.12.0
  - files_antivirus: 2.2.1
  - files_automatedtagging: 1.7.0
  - files_pdfviewer: 1.6.0
  - files_retention: 1.6.0
  - files_rightclick: 0.14.2
  - files_sharing: 1.9.0
  - files_trashbin: 1.7.0
  - files_versions: 1.10.0
  - files_videoplayer: 1.6.0
  - firstrunwizard: 2.6.0
  - gallery: 18.4.0
  - groupfolders: 5.0.5
  - logreader: 2.2.0
  - lookup_server_connector: 1.5.0
  - nextcloud_announcements: 1.6.0
  - notifications: 2.5.0
  - oauth2: 1.5.0
  - password_policy: 1.7.0
  - privacy: 1.1.0
  - provisioning_api: 1.7.0
  - recommendations: 0.5.0
  - richdocuments: 3.5.2
  - serverinfo: 1.7.0
  - sharebymail: 1.7.0
  - spreed: 7.0.2
  - support: 1.0.2
  - survey_client: 1.5.0
  - systemtags: 1.7.0
  - text: 1.1.1
  - theming: 1.8.0
  - twofactor_backupcodes: 1.6.0
  - updatenotification: 1.7.0
  - user_ldap: 1.7.0
  - viewer: 1.2.0
  - workflowengine: 1.7.0
Disabled:
  - admin_audit
  - encryption
  - files_external

**Nextcloud configuration:**
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.example.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "17.0.7.1",
        "overwrite.cli.url": "https:\/\/cloud.example.com",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "htaccess.RewriteBase": "\/",
        "logfile": "\/var\/log\/apache2\/cloud.example.com\/nextcloud.log",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***"
        },
        "filelocking.enabled": true,
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "maintenance": false,
        "loglevel": 2,
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}

**Are you using external storage, if yes which one:** none

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory

#### LDAP configuration (delete this part if not used)
+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Configuration                 | s01                                                                                                                                                                                |
+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | 1                                                                                                                                                                                  |
| homeFolderNamingRule          |                                                                                                                                                                                    |
| lastJpegPhotoLookup           | 0                                                                                                                                                                                  |
| ldapAgentName                 | ldapsearch@domain.example.com                                                                                                                                                       |
| ldapAgentPassword             | ***                                                                                                                                                                                |
| ldapAttributesForGroupSearch  |                                                                                                                                                                                    |
| ldapAttributesForUserSearch   |                                                                                                                                                                                    |
| ldapBackupHost                |                                                                                                                                                                                    |
| ldapBackupPort                |                                                                                                                                                                                    |
| ldapBase                      | dc=domain,dc=example,dc=com                                                                                                                                                         |
| ldapBaseGroups                | OU=gestman,dc=domain,dc=example,dc=com                                                                                                                                              |
| ldapBaseUsers                 | OU=gestman,dc=domain,dc=example,dc=com                                                                                                                                              |
| ldapCacheTTL                  | 600                                                                                                                                                                                |
| ldapConfigurationActive       | 1                                                                                                                                                                                  |
| ldapDefaultPPolicyDN          |                                                                                                                                                                                    |
| ldapDynamicGroupMemberURL     |                                                                                                                                                                                    |
| ldapEmailAttribute            | mail                                                                                                                                                                               |
| ldapExperiencedAdmin          | 0                                                                                                                                                                                  |
| ldapExpertUUIDGroupAttr       |                                                                                                                                                                                    |
| ldapExpertUUIDUserAttr        |                                                                                                                                                                                    |
| ldapExpertUsernameAttr        |                                                                                                                                                                                    |
| ldapExtStorageHomeAttribute   |                                                                                                                                                                                    |
| ldapGidNumber                 | gidNumber                                                                                                                                                                          |
| ldapGroupDisplayName          | cn                                                                                                                                                                                 |
| ldapGroupFilter               | (&(|(objectclass=group)))                                                                                                                                                          |
| ldapGroupFilterGroups         |                                                                                                                                                                                    |
| ldapGroupFilterMode           | 0                                                                                                                                                                                  |
| ldapGroupFilterObjectclass    | group                                                                                                                                                                              |
| ldapGroupMemberAssocAttr      | member                                                                                                                                                                             |
| ldapHost                      | srvdc.domain.gestman.lu                                                                                                                                                            |
| ldapIgnoreNamingRules         |                                                                                                                                                                                    |
| ldapLoginFilter               | (&(&(|(objectclass=person))(|(|(memberof:1.2.840.113556.1.4.1941:=CN=NextCloudUsers,OU=Groups,OU=GroupsContainerName,DC=domain,DC=example,DC=com)(primaryGroupID=4649))))(samaccountname=%uid)) |
| ldapLoginFilterAttributes     |                                                                                                                                                                                    |
| ldapLoginFilterEmail          | 0                                                                                                                                                                                  |
| ldapLoginFilterMode           | 1                                                                                                                                                                                  |
| ldapLoginFilterUsername       | 1                                                                                                                                                                                  |
| ldapNestedGroups              | 1                                                                                                                                                                                  |
| ldapOverrideMainServer        |                                                                                                                                                                                    |
| ldapPagingSize                | 500                                                                                                                                                                                |
| ldapPort                      | 389                                                                                                                                                                                |
| ldapQuotaAttribute            |                                                                                                                                                                                    |
| ldapQuotaDefault              |                                                                                                                                                                                    |
| ldapTLS                       | 0                                                                                                                                                                                  |
| ldapUserAvatarRule            | default                                                                                                                                                                            |
| ldapUserDisplayName           | displayname                                                                                                                                                                        |
| ldapUserDisplayName2          |                                                                                                                                                                                    |
| ldapUserFilter                | (&(|(objectclass=person))(|(|(memberof:1.2.840.113556.1.4.1941:=CN=NextCloudUsers,OU=Groups,OU=GroupsContainerName,DC=domain,DC=example,DC=com)(primaryGroupID=4649))))                         |
| ldapUserFilterGroups          | NextCloudUsers                                                                                                                                                                     |
| ldapUserFilterMode            | 1                                                                                                                                                                                  |
| ldapUserFilterObjectclass     | person                                                                                                                                                                             |
| ldapUuidGroupAttribute        | auto                                                                                                                                                                               |
| ldapUuidUserAttribute         | auto                                                                                                                                                                               |
| turnOffCertCheck              | 0                                                                                                                                                                                  |
| turnOnPasswordChange          | 0                                                                                                                                                                                  |
| useMemberOfToDetectMembership | 1                                                                                                                                                                                  |
+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

### Client configuration
**Browser:** curl

**Operating system:** Debian GNU/Linux

### Logs

Nothing relevant

#### Web server error log

Nothing relevant

#### Nextcloud log (data/nextcloud.log)

Example log entry from before I upgraded to 17.0.7

{"reqId":"P6fO6syiA1u831kImOwx","level":2,"time":"2020-06-07T07:24:54+00:00","remoteAddr":"A.B.C.D","user":"--","app":"user_ldap","method":"PROPFIND","url":"\/remote.php\/dav\/files\/path","message":"Bind failed: 49: Invalid credentials","userAgent":"curl\/7.64.0","version":"17.0.3.1"}

#### Browser log

Not applicable

### PR DESCRIPTION
**Pull request by lmamane**
Fix #21285

### CODE REVIEW

**Comment by lmamane** [comment]
This fixes issue #21285 but I cannot figure out how to add it to linked issue. I followed the github documentation at https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue, but:
1. "Linked issues" is not clickable, the doc says to click there.
1. The "fix #21285" in the description didn't do it either.

**Comment by kesselb** [comment]
Thanks for debugging :+1: 

Change makes sense to me. We need some changes for the tests. I will have a look later.

File: lib/private/Authentication/LoginCredentials/Store.php [hunk 1]
```diff
@@ -112,7 +112,7 @@ public function getLoginCredentials(): ICredentials {
 
 		if ($trySession && $this->session->exists('login_credentials')) {
 			$creds = json_decode($this->session->get('login_credentials'));
-			return new Credentials($creds->uid, $creds->uid, $creds->password);
+			return new Credentials($creds->uid, $this->session->get('loginname'), $creds->password);
```
Comment (ChristophWurst) [inline lib/private/Authentication/LoginCredentials/Store.php:115]: This is a bit of a red flag for me as this `loginname` value from the session is set elsewhere. So this relies on a few assumptions, making the code harder to understand and introducing undocumented dependencies between this code and the code that sets the value.

However, I see that we also use this same value in other places. That does not excuse the anti pattern, but it feels a bit less dirty.

In any case, can we maybe add a fallback to the UID when the session value is not there? Like ``$this->session->get('loginname') ?? $this->uid``?

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Fine from a functionality PoV, but a bit problematic in terms of code quality

Let's make it a bit less error-prone at least

**Comment by lmamane** [comment]
On Mon, Jun 08, 2020 at 12:01:04AM -0700, Christoph Wurst wrote:
> @ChristophWurst requested changes on this pull request.

> Fine from a functionality PoV, but a bit problematic in terms of code quality

I had started a bigger change, that would have added the loginname to
the associative array (key-value pair set) returned by
$this->session->get('login_credentials')
This entailed changing the event that sets login_credentials, namely
('OC_User', 'post_login'), and also ('\OC\User', 'postLogin') to add
this parameter. However, I wasn't sure if that would touch a
guaranteed stable public API (e.g. for apps) or not.

Changing the ('OC_User', 'post_login') looked rather safe, since it
gives an associative array; changing ('\OC\User', 'postLogin') looked
non-backwards compatible (since it gives multiple parameters, not a
single associative array) _but_ it looked like it could be a private
internal API, and changing all listeners within Nextcloud itself would
be sufficient? But I wasn't sure.

Then I saw this code pattern
$this->session->get('loginname')
elsewhere in the code and thought "ah, that's how they do it".

If you tell me my first idea was better, and that the "public API"
worries are unfounded, I will gladly submit a new pull request based
in that.

Please let me know.

> > @@ -112,7 +112,7 @@ public function getLoginCredentials(): ICredentials {
>
>  		if ($trySession && $this->session->exists('login_credentials')) {
>  			$creds = json_decode($this->session->get('login_credentials'));
> -			return new Credentials($creds->uid, $creds->uid, $creds->password);
> +			return new Credentials($creds->uid, $this->session->get('loginname'), $creds->password);

> This is a bit of a red flag for me as this `loginname` value from
> the session is set elsewhere. So this relies on a few assumptions,
> making the code harder to understand and introducing undocumented
> dependencies between this code and the code that sets the value.

> In any case, can we maybe add a fallback to the UID when the session
> value is not there? Like ``$this->session->get('loginname') ??
> $this->uid``?

I would not put a fallback to a value that is a wrong one, and
systematically so in some scenarios (such as LDAP external
authentication with uid != loginname). That already created a subtle
and hard to track bug (app tokens would work for a few minutes and
then silently stop working...), I don't think keeping such things in
the codebase is a good idea. If we cannot get the loginname, then we
can return an _exception_, or maybe the value null?

**Comment by kesselb** [comment]
> Then I saw this code pattern
$this->session->get('loginname')

https://github.com/nextcloud/server/blame/caff1023ea72bb2ea94130e18a2a6e2ccf819e5f/apps/encryption/lib/Controller/SettingsController.php#L116-L120

Seems the encryption app run into a similar issue before. Adding the loginName to the event sounds good but is quite hard to backport.

**Comment by lmamane** [comment]
On the "hard to backport" part, we could imagine having this one-liner fix in the stable branches, and the event change in master?

**Comment by ChristophWurst** [comment]
> On the "hard to backport" part, we could imagine having this one-liner fix in the stable branches, and the event change in master?

Fine by me

File: lib/private/Authentication/LoginCredentials/Store.php (hunk 1 above)
Comment (lmamane) [reply lib/private/Authentication/LoginCredentials/Store.php]: (addresses in main thread)

**Comment by lmamane** [comment]
> > On the "hard to backport" part, we could imagine having this one-liner fix in the stable branches, and the event change in master?
> Fine by me

My current patch backports easily to stable19 branch. After you are happy with the current patch, let me know what version you want on stable19 (event change or "dirty" one-liner).

File: lib/public/User/Events/PostLoginEvent.php [hunk 2]
```diff
@@ -61,6 +68,13 @@ public function getUser(): IUser {
 		return $this->user;
 	}
 
+	/**
+	 * @since 20.0.0
+	 */
+	public function getLoginName(): string {
+		return $this->login_name;
```
Comment (kesselb) [inline lib/public/User/Events/PostLoginEvent.php:75]:
```suggestion
		return $this->loginName;
```
?

File: lib/private/Server.php [hunk 3]
```diff
@@ -557,9 +557,9 @@ public function __construct($webRoot, \OC\Config $config) {
 				$dispatcher = $this->query(IEventDispatcher::class);
 				$dispatcher->dispatchTyped(new BeforeUserLoggedInEvent($uid, $password));
 			});
-			$userSession->listen('\OC\User', 'postLogin', function ($user, $password, $isTokenLogin) {
+			$userSession->listen('\OC\User', 'postLogin', function ($user, $login_name, $password, $isTokenLogin) {
```
Comment (kesselb) [inline lib/private/Server.php:560]:
```suggestion
			$userSession->listen('\OC\User', 'postLogin', function ($user, $loginName, $password, $isTokenLogin) {
```

File: lib/private/Server.php [hunk 4]
```diff
@@ -557,9 +557,9 @@ public function __construct($webRoot, \OC\Config $config) {
...
 				/** @var $user \OC\User\User */
-				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'password' => $password, 'isTokenLogin' => $isTokenLogin]);
+				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'login_name' => $login_name, 'password' => $password, 'isTokenLogin' => $isTokenLogin]);
```
Comment (kesselb) [inline lib/private/Server.php:562]:
```suggestion
				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'loginName' => $loginName, 'password' => $password, 'isTokenLogin' => $isTokenLogin]);
```

File: lib/private/Authentication/LoginCredentials/Store.php [hunk 5]
```diff
@@ -112,7 +112,7 @@ public function getLoginCredentials(): ICredentials {
 
 		if ($trySession && $this->session->exists('login_credentials')) {
 			$creds = json_decode($this->session->get('login_credentials'));
-			return new Credentials($creds->uid, $creds->uid, $creds->password);
+			return new Credentials($creds->uid, $creds->login_name, $creds->password);
```
Comment (kesselb) [inline lib/private/Authentication/LoginCredentials/Store.php:115]:
```suggestion
			return new Credentials($creds->uid, $creds->loginName, $creds->password);
```

**Review by kesselb — CHANGES_REQUESTED** [review]
camelCase ;)

File: lib/public/User/Events/PostLoginEvent.php [hunk 6]
```diff
@@ -47,9 +53,10 @@ class PostLoginEvent extends Event {
 	/**
 	 * @since 18.0.0
 	 */
-	public function __construct(IUser $user, string $password, bool $isTokenLogin) {
+	public function __construct(IUser $user, string $loginName, string $password, bool $isTokenLogin) {
```
Comment (blizzz) [inline lib/public/User/Events/PostLoginEvent.php:56]: FYI, we cannot backport this as it breaks public API.

**Comment by MorrisJobke** [comment]
> My current patch backports easily to stable19 branch. After you are happy with the current patch, let me know what version you want on stable19 (event change or "dirty" one-liner).

The one liner.

**Comment by MorrisJobke** [comment]
> > My current patch backports easily to stable19 branch. After you are happy with the current patch, let me know what version you want on stable19 (event change or "dirty" one-liner).
> 
> The one liner.

Stable19: #21779

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by MorrisJobke** [comment]
Rebased to fix a conflict.

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**